Analyze nanopore data to see which phagemid lines are monoclonal


```{r}
library(tidyverse)
library(here)
library(fs)
library(furrr)
plan(multicore, workers = 3)
import::from(Biostrings, complement, reverseComplement, width, DNAString,
  DNAStringSet, readDNAStringSet, subseq,
  vmatchPDict, vcountPDict
)
import::from(DECIPHER, TrimDNA, AlignSeqs, OrientNucleotides, BrowseSeqs)

```

```{r}
seq_barcodes <- here('output', '2022-05-18-sequencing-barcodes',
  'sequencing-barcodes.csv') %>%
  read_csv %>%
  glimpse
```

Forward and reverse primers for the sequencing barcodes

```{r}
fwd <- seq_barcodes %>% pull(primer_1) %>% unique
rev <- seq_barcodes %>% pull(primer_2) %>% unique
rev_rc <- rev %>% DNAString %>% reverseComplement %>% as.character
stopifnot(identical(length(fwd), 1L))
stopifnot(identical(length(rev), 1L))
```

```{r}
eblocks <- here('analysis/2022-05-18-sequencing-barcodes',
  '2022-05-20-eblock-template-mm.csv') %>% read_csv
```


# Improved pipeline - Combo approach

First trim by primers (try both orientations and pick best), then match.

#. Combine all sequences into a single DNAStringSet
#. Give a common orientation
#. check that this is the correct oreitnation for the quant barcodes; if not, RC it.
#. Trim using the primers
#. Do some QC - check stats of matches; check that see the spacer and probe sequences
#. For successfully trimmed seqs, match against barcodes
#. Use to assess single or multiple copies (with some confidence score, based on how many total reads passing QC
#. Of ones with single q-seq barcode, check that they have the correct qPCR
#barcode
#. pick 25 with desired 5X5 design

Should think of some QC to do within each of these steps.

Should also consider if/how to do this for the qPCR barcodes, to check that the 

## Combine all sequences into a single DNAStringSet

```{r}
fns <- here('data', 
  '2022-06-03-quantification-plasmid-nanopore-sequencing/reads') %>%
  dir_ls(recurse = TRUE, glob = '*.fastq')

fns %>%
  str_subset('unclassified', negate = TRUE) %>%
  length
#> reads <- readDNAStringSet(fns, format = 'fastq')
```

Trying to read all the files in with one call to readDNAStringSet leads to an error, I think because of too many open connections.

So instead read each folder in one by one, then join into one DNAStringSet

```{r}
dirs <- here('data',
  '2022-06-03-quantification-plasmid-nanopore-sequencing/reads') %>%
  dir_ls
reads_list <- dirs %>%
  map(dir_ls, glob = '*.fastq') %>%
  map(readDNAStringSet, format = 'fastq')
reads_raw <- reads_list %>% reduce(c)
# Drop unclassified reads
reads <- reads_raw %>%
  {.[str_detect(names(.), 'unclassified', negate = TRUE)]}
```

Note, the indexing barcode is reported at the end of the read name.

```{r}
names(reads) %>% str_sub(start = - 14) %>% sample(3)
```

## Orient all reads to a common reference orientation

First, let's add a reference sequence that we can use to set the orientation.

```{r}
ref <- eblocks %>% slice(1) %>% pull(sequence) %>% 
  set_names('reference') %>%
  DNAStringSet
reads_with_ref <- c(ref, reads)
```

```{r}
reads_oriented <- OrientNucleotides(reads_with_ref, reference = 1, processors = 3)
```

TODO: concider the threshold and orientation settings

## Trim using the primers

```{r}
reads_oriented_trimmed <- TrimDNA(reads_oriented, fwd, rev_rc, 
  #> minWidth = 10,
  type = 'sequences')
```

Does it seem to be working?

```{r}
probe_length <- seq_barcodes$probe_length %>% unique
ont_barcode_length <- seq_barcodes$ont_barcode_length %>% unique
expected_length <- 4*3 + ont_barcode_length + probe_length
```

```{r}
sum(width(reads_oriented_trimmed) < width(reads_oriented))
mean(width(reads_oriented_trimmed) < width(reads_oriented))
width(reads_oriented) %>% summary
width(reads_oriented_trimmed) %>% summary

mean(width(reads_oriented_trimmed) == expected_length)
mean(width(reads_oriented_trimmed) < 100)

width(reads_oriented_trimmed) %>% qplot +
  scale_x_log10() +
  expand_limits(x = 10) +
  geom_vline(xintercept = expected_length, color = 'darkred')
```

Looks like it worked; though most of the reads were not successfully trimmed at both primers.

TODO: consider TrimDNA options and how might increase trimming.

```{r}
reads_post <- reads_oriented_trimmed %>%
  {.[width(.) < 100]}
```

## Do some QC - check stats of matches; check that see the spacer and probe sequences

TODO

## For successfully trimmed seqs, match against barcodes

```{r}
```

```{r}
ont_barcodes <- seq_barcodes %>% 
  select(ont_barcode_id, ont_barcode_sequence) %>%
  deframe %>%
  DNAStringSet
```

TODO: Redo with less strict matching (will be slower)

```{r}
xmat <- vcountPDict(ont_barcodes, reads_post, )
rownames(xmat) <- names(ont_barcodes)
colnames(xmat) <- names(reads_post)
#> ~vcountPDict(pats, .x, with.indels = TRUE, max.mismatch = 2)),
```

possible improvements

- get the best match, up to sufficient closeness
<!--  -->

Check that only one match per read:

```{r}
stopifnot( (colSums(xmat) %>% max) == 1 )
```

What fraction of reads match any barcode?

```{r}
(colSums(xmat) >= 1) %>% mean %>% round(2)
```

- Turn into a dataframe with one row per read, filtered to matches
  - for qc, might want to track the total number of reads / non-matches
- extract the indexing barcode from the read name for well-matching
<!--  -->

```{r}
x <- xmat %>%
  t %>%
  as_tibble(rownames = 'read_name') %>%
  pivot_longer(cols = starts_with('BC'), names_to = 'ont_barcode', values_to = 'match') %>%
  filter(match == 1) %>%
  mutate(
    index_barcode = str_extract(read_name, '(?<=barcode=)bw_[0-9]{3}$')
  )
```


```{r}
x %>%
  filter(read_name != 'reference') %>%
  with_groups(index_barcode, count, ont_barcode)
```

Notice that there are many cases where a library has one or a few hits to a minority barcode.
It is worth checking what is going on here.

## Use to assess single or multiple copies (with some confidence score, based on how many total reads passing QC


```{r}
x1 <- x %>%
  filter(read_name != 'reference') %>%
  with_groups(index_barcode, count, ont_barcode) %>%
  with_groups(index_barcode, mutate,
    frac = n / sum(n)
  )
x1 %>%
  filter(frac > 0.05) %>%
  count(index_barcode) %>%
  pull(n) %>%
  table
```


## Of ones with single q-seq barcode, check that they have the correct qPCR barcode

## pick 25 with desired 5X5 design

