

# Setup

```{r}
library(here)
library(tidyverse)
library(fs)
library(googlesheets4)
# library(speedyseq)
# library(metacal)

# library(cowplot)
# theme_set(theme_cowplot())
# library(patchwork)
# library(ggbeeswarm)

import::from(Biostrings, readDNAStringSet, DNAString, DNAStringSet,
  writeXStringSet, reverseComplement)
```

Download the ONT barcodes from Google Sheets, and save in local repo (do once).

```{r, eval = F}
sheet_url <- 'https://docs.google.com/spreadsheets/d/1OLUgNzy3n6Gc0TSpBiqpHm6B2WPCaBOYvrJCqT2Iplo'
ont_barcodes <- read_sheet(sheet_url, 'barcodes', 
  col_names = c('id', 'sequence', 'blank', 'length')
  ) %>%
  glimpse
ont_barcodes <- ont_barcodes %>%
  select(id, sequence) %>%
  mutate(length = nchar(sequence))
write_csv(ont_barcodes, here('data/oxford-nanopore-barcodes.csv'))
```

Load ONT barcodes from local copy, dropping 'BC12A'.

```{r}
ont_barcodes <- read_csv(here('data/oxford-nanopore-barcodes.csv'), 
  col_types = 'cci') %>%
  rename_with(~str_c('ont_barcode_', .)) %>%
  filter(ont_barcode_id != 'BC12A') %>%
  glimpse
stopifnot(all(ont_barcodes$ont_barcode_length == 24))
```


```{r}
fns <- c(
  'output/2022-05-02-barcode-primers/barcodes-postprocessed.csv',
  'output/2022-05-05-barcode-primers/barcodes-postprocessed.csv'
) %>%
  here
qpcr_barcodes <- tibble(set = 1:2, file = fns) %>%
  mutate(
    data = map(file, read_csv, col_types = 'cicicicici')
  ) %>%
  unnest(data) %>%
  mutate(
    primer_2_rc = map_chr(primer_2,
      ~ DNAString(.x) %>% reverseComplement %>% as.character)
  ) %>%
  glimpse
```

# Generate candidate amplicons

Choose a model set of primers, probe, and spacers,
We'll use Set 1 Replicate 9 based on its good R2 and 98% efficiency.
I'll pull spacers 2 and 3 from it as well, to maintain the same context around the probe.
Spacer 1 is just a random 4 base string with 50% GC.

```{r}
model <- qpcr_barcodes %>%
  filter(set == 1, replicate == 9) %>%
  select(starts_with(c('primer', 'probe'))) %>%
  mutate(
    spacer_1 = 'AGTC',
    spacer_2 = 'TGAC',
    spacer_3 = 'TCTA',
  ) %>%
  glimpse
```

Form the cross with all ONT barcodes,
(When assembling, remember to use the reverse complement of the reverse primer.)

```{r}
seq_barcodes <- crossing(model, ont_barcodes) %>%
  mutate(
    amplicon = str_c(
      primer_1, 
      spacer_1, 
      ont_barcode_sequence, 
      spacer_2, 
      probe,
      spacer_3, 
      primer_2_rc),
    amplicon_length = nchar(amplicon)
  )
```


```{r}
seq_barcodes %>% pull(amplicon) %>% head 
```

To use the batch mode in IDT Analyzer (https://www.idtdna.com/calc/analyzer/home/batch), let's export a fasta file,

```{r}
```


```{r}
out_dir <- here('output', '2022-05-18-sequencing-barcodes')
dir_create(out_dir)
fn <- path(out_dir, 'candidate-amplicons.fasta')

amplicons_dna <- seq_barcodes %>%
  transmute(
    sequence_id = str_glue('{ont_barcode_id}_amplicon'),
    sequence = amplicon
  ) %>%
  deframe %>%
  DNAStringSet

writeXStringSet(amplicons_dna, fn)
```

Note, the melting temps look higher when I use the 'qPCR' parameter set.
Need to pay attention to what parameter set is used, and record this.


# Full tag sequences

Full tracer tag:
[Gibson overhang 1] - NNNN - [qPCR barcode] - [10 or 20 bp spacer] - [Sequencing Barcode] - NNNN - [Gibson overhang 2]


qPCR barcodes from set 2; pick the 5 with the lowest efficiency, and good R2:

7
19
10
4
18

```{r}
reps <- c(7, 19, 10, 4, 18)
qpcr_barcodes_picked <- qpcr_barcodes %>%
  filter(set == 2, replicate %in% reps)
```

```{r}
rc <- function(s) {
  s %>% DNAString %>% reverseComplement %>% as.character
}
```


5’ end: tctcattttcgccaaaagttgg (61C melt temp, 40.91% G/C)
3’ end: ctaggagggtcagccgttaag (62C melt temp, 58% G/C)


```{r}
set.seed(46)
l <- c('A', 'C', 'G', 'T')
spacers <- list(
  overhang_1 = 'tctcattttcgccaaaagttgg' %>% toupper,
  overhang_2 = 'ctaggagggtcagccgttaag' %>% toupper,
  tag_spacer_1 = sample(l, size = 4, replace = TRUE),
  tag_spacer_2 = sample(l, size = 10, replace = TRUE),
  tag_spacer_3 = sample(l, size = 4, replace = TRUE)
) %>%
  map_chr(str_c, collapse='') %>%
  print
tags_spacers <- spacers %>%
  enframe() %>%
  pivot_wider() %>%
  mutate(
    overhang_2_rc = rc(overhang_2),
  )
```

```{r}
tags_qpcr <- qpcr_barcodes_picked %>%
  transmute(
    qpcr_barcode = barcode,
    qpcr_barcode_set = set,
    qpcr_barcode_replicate = replicate,
    qpcr_barcode_replicate_string = formatC(qpcr_barcode_replicate, width = 3, flag = "0"),
    qpcr_barcode_idx = row_number(),
  )
tags_seq <- seq_barcodes %>%
  transmute(
    seq_barcode = amplicon, 
    ont_barcode_id, 
    seq_barcode_idx = row_number(),
    #> here - how to pair up specific seq barcodes with specific qpcr barcodes?
    qpcr_barcode_idx = rep_len(1:5, length.out = 96)
  )
```


```{r}
```

names:

`T1_QS2_###_BC##`


T1 = Tag Set 1
QS2 = qPCR barcode set 2
BC = ON barcode


```{r}
tags <- tags_seq %>%
  left_join(tags_qpcr, by = 'qpcr_barcode_idx') %>%
  crossing(tags_spacers) %>%
  mutate(
    tag_sequence = str_c(
      overhang_1,
      tag_spacer_1, 
      qpcr_barcode,
      tag_spacer_2, 
      seq_barcode,
      tag_spacer_3,
      overhang_2 # NOTE: Don't reverse complement
      ),
    tag_length = nchar(tag_sequence),
    tag_name = str_glue(
      'T1_QS{qpcr_barcode_set}_{qpcr_barcode_replicate_string}_{ont_barcode_id}'
    )
  )
tags$tag_length %>% summary
```


```{r}
write_csv(tags, 'tags.csv')

tags_dna <- tags %>%
  select(tag_name, tag_sequence) %>%
  deframe %>%
  DNAStringSet

writeXStringSet(tags_dna, 'tags.fasta')
```

# eblock test template

```{r}

```

Ah ok, I think I understand - you . Let me programmatically confirm the sequences in your excel file are identical to the oligo seq’s except for this insertion, should take <10m

The eBlock's need to be >= 300bp. To reach this length, AG inserted GTCA after CCAT (the first tag spacer).

Let's confirm that the sequences that AG created and shared are identical to what I get if I insert this additional sequence.

The seq's AG created:

```{r}
eblocks_ag <- readxl::read_excel('eblock-test-template.xlsx')
eblocks_ag%>%
  filter(!is.na(...12)) %>%
  glimpse
eblocks_ag <- eblocks_ag %>%
  select(1:3) %>%
  janitor::clean_names() %>%
  mutate(length = nchar(sequence))
eblocks_ag_dna <- eblocks_ag %>%
  select(name, sequence) %>%
  deframe %>%
  DNAStringSet
```

My version:

```{r}
eblocks_mm <- tags %>%
  mutate(
    tag_spacer_1_eblock = str_c(tag_spacer_1, 'GTCA'),
    tag_sequence = str_c(
      overhang_1,
      tag_spacer_1_eblock, 
      qpcr_barcode,
      tag_spacer_2, 
      seq_barcode,
      tag_spacer_3,
      overhang_2 # NOTE: Don't reverse complement
      ),
    tag_length = nchar(tag_sequence),
  )
eblocks_mm$tag_length %>% summary
eblocks_mm_dna <- eblocks_mm %>%
  select(tag_name, tag_sequence) %>%
  deframe %>%
  DNAStringSet
```

Confirm that the order is the same,

```{r}
stopifnot(all.equal(eblocks_ag_dna %>% names, eblocks_mm_dna %>% names))
```

Check for differences,

```{r}
all.equal(eblocks_ag_dna %>% as.character, eblocks_mm_dna %>% as.character)
eblocks_ag_dna[eblocks_ag_dna != eblocks_mm_dna] %>% names
```

Differences in 13 sequences, need to check these.

```{r, eval = F}
a <- eblocks_ag_dna['T1_QS2_018_BC14']
b <- eblocks_mm_dna['T1_QS2_018_BC14']
names(a) <- str_c(names(a), '_ag')
names(b) <- str_c(names(b), '_mm')
x <- c(a,b)
x_aligned <- DECIPHER::AlignSeqs(x)
DECIPHER::BrowseSeqs(x_aligned, highlight=0)
```

Looks like AG accidentally added extra GTCA's, possibly after every CCAT instead of just the first one.

Can check that the differences in the seq's are alwaya 4bp*X too long; or that only the sequences with extra CCAT's are the issue.

## Additional check

```{r}
a <- str_c(
  str_sub(tags$tag_sequence, 1, 26),
  'GTCA',
  str_sub(tags$tag_sequence, 27)
)
b <- eblocks_mm %>% pull(tag_sequence)
identical(a,b)
#> [1] TRUE
```

## Save the correct sequences

```{r}
stopifnot(identical(eblocks_ag$name, eblocks_mm$tag_name %>% as.character))
eblocks <- eblocks_mm %>%
  add_column(well_position = eblocks_ag$well_position) %>%
  select(well_position, name = tag_name, sequence = tag_sequence)
write_csv(eblocks, '2022-05-20-eblock-template-mm.csv')
```



